modules/hostap/src/supp_main.c

```diff
@@ -1079,12 +1079,6 @@ static void zephyr_hostapd_init(struct supplicant_context *ctx)
 		zephyr_hostapd_ctrl_init((void *)interfaces->iface[i]->bss[0]);
 	}
 
-	ret = wifi_nm_register_mgd_iface(wifi_nm_get_instance("hostapd"), iface);
```

Comment:
The driver only registers the interface and then it's up to higher level stack to decide how to manage it, with or without NM. And interface registration has to be done by the NM for all the interfaces it manages not the drivers.

Reply:
But we should call `wifi_nm_register_mgd_iface` for SAP first, then can support `net_if_get_wifi_sap`, right?
`Line 1044` in this file already calls `net_if_get_wifi_sap`.
So add `wifi_nm_register_mgd_iface(wifi_nm_get_instance("hostapd"), iface);` here is wrong logic.

Reply:
Yes, but how can the hostapd/supplicant layer know which interface should be SAP or STA? How to decide it?

Reply:
We cannot, it's the user that gets to decide :).

Reply:
Hi @krish2718 any suggestions? where should we call wifi_nm_register_mgd_iface for SAP?

Reply:
We don't have a mechanism to take the user input for interface -> type mapping e..g, wlan0->STA, wlan1->SAP, wlan2->AP etc. From top of my head we can use the DTS for this (There's an ongoing [debate](https://github.com/zephyrproject-rtos/zephyr/issues/77638) on using DTS for SW/Network config though), e.g., `hostap_ifaces` with `compatible=hostap,sta` -> STA's and `compatible=hostap,ap` as AP and then supplicant will use these and do the corresponding registration.

```
hostap_iface0: wlan0 {
   compatible="hostap,sta";
}

hostap_iface1: wlan1 {
   compatible="hostap,ap";
}
```

This is just a dirty tree to show my thoughts.

@jukkar @rlubos WDYT? If DTS is not acceptable as a SW configuration then alternative is to use `net_if_set_name` and then let supplicant add a Kconfig with a list of strings for STA and SAP.

E.g.,

````
config WIFI_NM_SUPPLICANT_STA_IFACES
   string "STA iface globs"
   default "wlan*"

config WIFI_NM_SUPPLICANT_AP_IFACES
   string "AP iface globs"
   default "ap*"
```

And user's can then use `net_if_set_name` to rename the network interfaces and supplicant register them accordingly.

Reply:
Hi @krish2718 

> to use `net_if_set_name` and then let supplicant add a Kconfig with a list of strings for STA and SAP.

Hi @krish2718 still the same question, during the hostapd init, how to get the net iface for SAP? You mean add one function named like `net_if_get_by_name_prefix` to get the net iface with name "ap_xxx"?

Reply:
Note that network interface name is not a good candidate for selecting a network interface because the name can be changed at runtime and can be whatever application wants.

Reply:
> hat network interface name is not a good candidate for selecting a network interface because the name can be changed at runtime and can be whatever application wants.

Sure, hence we need a clear user input either DTS or YAML or whatever network configuration mechanism. This is just a workaround based on some guidelines.

> Hi @krish2718 still the same question, during the hostapd init, how to get the net iface for SAP? You mean add one function named like net_if_get_by_name_prefix to get the net iface with name "ap_xxx"?

I would just extend https://github.com/zephyrproject-rtos/zephyr/blob/main/modules/hostap/src/supp_main.c#L245 and then do the NM iface registration here https://github.com/zephyrproject-rtos/zephyr/blob/main/modules/hostap/src/supp_main.c#L503 based on Either DTS compatible or just interface name.

Reply:
Sure, can you please open a GH issue with these details?

Reply:
Thanks, please see this issue:
https://github.com/zephyrproject-rtos/zephyr/issues/83087